javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -334,11 +348,7 @@ default int leafCount() {
      * @return The number of nodes of this tree.
      */
     default int nodeCount() {
-        if (isEmpty()) {
-            return 0;
-        } else {
-            return 1 + getChildren().foldLeft(0, (count, child) -> count + child.nodeCount());
-        }
+        return length();
```

Comment:
> So you're simply using it for multiple implementation inheritance (traits style), right?

Yes. But there is one more reason: High cohesion. It is nice to see the behavior of a function (/method) in one place. In our example we have just two distinct impls (Empty and Node) but there are situations where we might have even more (e.g. Root, Inner, Leaf). Spreading the behavior of one method over the classes makes an algorithm hard to read.

An Interface/Trait is like a blueprint of a type, we see how it basically works. By looking at the impls we see the specific characteristics. Those are often one-pagers.